File: advent_15_II.py
import sys
from collections import deque

sys.setrecursionlimit(10 ** 6)

# Definitions for movements
dr = [0, 0, 1, -1]
dc = [1, -1, 0, 0]


def findbox(a, i, j, b, d):
    n = len(a)
    m = len(a[0])
    visit = [[False for _ in range(m)] for _ in range(n)]
    q = deque()
    comp = []
    q.append((i, j))
    visit[i][j] = True

    while q:
        cur = q.popleft()
        r, c = cur
        comp.append(((r, c), a[r][c]))
        if a[r][c] == '[' and not visit[r][c + 1]:
            visit[r][c + 1] = True
            q.append((r, c + 1))
        elif a[r][c] == ']' and not visit[r][c - 1]:
            visit[r][c - 1] = True
            q.append((r, c - 1))

        nr, nc = r + b, c + d
        if (a[nr][nc] == '[' or a[nr][nc] == ']') and not visit[nr][nc]:
            visit[nr][nc] = True
            q.append((nr, nc))

    return comp


def move(a, comp, di, dj):
    for it in comp:
        r, c = it[0]
        if a[r + di][c + dj] == '#':
            return False
    for it in comp:
        r, c = it[0]
        a[r][c] = '.'
    for it in comp:
        r, c = it[0]
        a[r + di][c + dj] = it[1]
    return True


def findpar(node, par):
    if node == par[node]:
        return node
    par[node] = findpar(par[node], par)
    return par[node]


def ubs(u, v, par, size):
    ulp_u = findpar(u, par)
    ulp_v = findpar(v, par)
    if ulp_v == ulp_u:
        return
    if size[ulp_u] < size[ulp_v]:
        par[ulp_u] = ulp_v
        size[ulp_v] += size[ulp_u]
    else:
        par[ulp_v] = ulp_u
        size[ulp_u] += size[ulp_v]


def main():
    n, m = 50, 50
    a = [['.' for _ in range(2 * m)] for _ in range(n)]

    for i in range(n):
        temp = input().strip()
        for j in range(m):
            if temp[j] == '#':
                a[i][2 * j] = '#'
                a[i][2 * j + 1] = '#'
            elif temp[j] == '.':
                a[i][2 * j] = '.'
                a[i][2 * j + 1] = '.'
            elif temp[j] == '@':
                a[i][2 * j] = '@'
                a[i][2 * j + 1] = '.'
            elif temp[j] == 'O':
                a[i][2 * j] = '['
                a[i][2 * j + 1] = ']'

    sz = 20000
    x, y = 20, 1000

    # Initialize the grid
    temp = []
    aesehi = input().strip()
    # Read the grid row by row
    for i in range(x):
        row = input().strip()
        if len(row) != y:
            raise ValueError(f"Row {i + 1} does not have exactly {y} characters.")
        temp.append(list(row))

    # Flatten the grid into a moves list
    moves = [temp[i][j] for i in range(x) for j in range(y)]

    for i in range(n):
        for j in range(m):
            print(a[i][2 * j] + a[i][2 * j + 1], end="")
        print()
    print()

    r, c = -1, -1
    for i in range(n):
        for j in range(2 * m):
            if a[i][j] == '@':
                r, c = i, j
                break

    for i in range(sz):
        if moves[i] == '^':
            if a[r - 1][c] == '.':
                a[r][c], a[r - 1][c] = a[r - 1][c], a[r][c]
                r -= 1
            elif a[r - 1][c] == '#':
                continue
            else:
                temp = findbox(a, r - 1, c, -1, 0)
                temp.append(((r, c), '@'))
                if move(a, temp, -1, 0):
                    r -= 1
        elif moves[i] == 'v':
            if a[r + 1][c] == '.':
                a[r][c], a[r + 1][c] = a[r + 1][c], a[r][c]
                r += 1
            elif a[r + 1][c] == '#':
                continue
            else:
                temp = findbox(a, r + 1, c, 1, 0)
                temp.append(((r, c), '@'))
                if move(a, temp, 1, 0):
                    r += 1
        elif moves[i] == '<':
            if a[r][c - 1] == '.':
                a[r][c], a[r][c - 1] = a[r][c - 1], a[r][c]
                c -= 1
            elif a[r][c - 1] == '#':
                continue
            else:
                j = c - 1
                while a[r][j] in [']', '[']:
                    j -= 1
                if a[r][j] == '#':
                    continue
                else:
                    for k in range(j, c):
                        a[r][k] = a[r][k + 1]
                    a[r][c] = '.'
                    c -= 1
        elif moves[i] == '>':
            if a[r][c + 1] == '.':
                a[r][c], a[r][c + 1] = a[r][c + 1], a[r][c]
                c += 1
            elif a[r][c + 1] == '#':
                continue
            else:
                j = c + 1
                while a[r][j] in ['[', ']']:
                    j += 1
                if a[r][j] == '#':
                    continue
                else:
                    for k in range(j, c, -1):
                        a[r][k] = a[r][k - 1]
                    a[r][c] = '.'
                    c += 1

        print(moves[i])
        print(i + 1)
        for x in range(n):
            for y in range(2 * m):
                print(a[x][y], end="")
            print()
        print()

    ans = 0
    for i in range(n):
        for j in range(2*m):
            if a[i][j] == '[':
                ans += 100 * i + j
    print(ans)


if __name__ == "__main__":
    main()

File: advent_15_II.cpp
#include<bits/stdc++.h>

using namespace std;

#define ll long long int
#define pb push_back
#define f first
#define s second
#define ld long double

vector<ll>dr={0,0,1,-1};
vector<ll>dc={1,-1,0,0};


vector<pair<pair<ll,ll>,char>> findbox(vector<vector<char>>&a,ll i, ll j, ll b,ll d){
    ll n=a.size();
    ll m=a[0].size();
    vector<vector<bool>>visit(n,vector<bool>(m,false));
    queue<pair<ll,ll>>q;
    vector<pair<pair<ll,ll>,char>>comp;
    q.push({i,j});
    visit[i][j]=true;
    while(!q.empty()){
        auto cur=q.front();
        
        q.pop();
        ll r=cur.first,c=cur.second;
        comp.pb({{r,c},a[r][c]});
        if(a[r][c]=='[' && !visit[r][c+1]){
            visit[r][c+1]=true;
            q.push({r,c+1});
        }
        else if(a[r][c]==']' && !visit[r][c-1]){
            visit[r][c-1]=true;
            q.push({r,c-1});
        }
        
            ll nr=r+b,nc=c+d;
            if((a[nr][nc]=='[' || a[nr][nc]==']') && !visit[nr][nc]){
                visit[nr][nc]=true;
                q.push({nr,nc});
            }
        
    }
    return comp;
}

bool move(vector<vector<char>>&a,vector<pair<pair<ll,ll>,char>>&comp,ll di,ll dj){
    for(auto it:comp){
        ll r=it.first.first,c=it.first.second;
        if(a[r+di][c+dj]=='#') return false;
    }
    for(auto it:comp){
        a[it.f.f][it.f.s]='.';
    }
    for(auto it:comp){
        a[it.f.f+di][it.f.s+dj]=it.s;
    }
    return true;
}

// code for disjoint set union (by size)
ll findpar(ll node,vector<ll>&par){
    if(node==par[node]) return node;
    return par[node]=findpar(par[node],par);
}
void ubs(ll u, ll v,vector<ll>&par,vector<ll>&size){
    ll ulp_u=findpar(u,par);
    ll ulp_v=findpar(v,par);
    if(ulp_v==ulp_u) return;
    if(size[ulp_u]<size[ulp_v]){
        par[ulp_u]=ulp_v;
        size[ulp_v]+=size[ulp_u];
    }
    else{
        par[ulp_v]=ulp_u;
        size[ulp_u]+=size[ulp_v];
    }
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    ll n,m;
    n=50,m=50;
    vector<vector<char>>a(n,vector<char>(2*m));
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            char temp;
            cin>>temp;
            if(temp=='#'){
                a[i][2*j]='#';
                a[i][2*j+1]='#';
            }
            else if(temp=='.'){
                a[i][2*j]='.';a[i][2*j+1]='.';
            }
            else if(temp=='@'){
                a[i][2*j]='@';a[i][2*j+1]='.';
            }
            else if(temp=='O'){
                a[i][2*j]='[';a[i][2*j+1]=']';
            }
        }
    }
    ll sz=20000;
    ll x=20,y=1000;
    vector<vector<char>>temp(x,vector<char>(y));;
    for(int i=0;i<x;i++) for(int j=0;j<y;j++) cin>>temp[i][j];
    vector<char>moves;
    for(int i=0;i<x;i++) for(int j=0;j<y;j++) moves.pb(temp[i][j]);
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++) cout<<a[i][2*j]<<a[i][2*j+1];
        cout<<endl;
    }
    cout<<endl;
    ll r,c;
    for(int i=0;i<n;i++){
        for(int j=0;j<2*m;j++){
            if(a[i][j]=='@'){
                r=i,c=j;break;
            }
        }
    }
    for(int i=0;i<sz;i++){
        if(moves[i]=='^'){
            if(a[r-1][c]=='.'){
                swap(a[r][c],a[r-1][c]);
                r--;
            }
            else if(a[r-1][c]=='#') continue;
            else{
                vector<pair<pair<ll,ll>,char>>temp=findbox(a,r-1,c,-1,0);
                temp.pb({{r,c},'@'});
                if(move(a,temp,-1,0)) r--;
            }
        }
        else if(moves[i]=='v'){
            if(a[r+1][c]=='.'){
                swap(a[r][c],a[r+1][c]);
                r++;
            }
            else if(a[r+1][c]=='#') continue;
            else{
                vector<pair<pair<ll,ll>,char>>temp=findbox(a,r+1,c,1,0);
                temp.pb({{r,c},'@'});
                if(move(a,temp,1,0)) r++;
            }
        }
        else if(moves[i]=='<'){
            if(a[r][c-1]=='.'){
                swap(a[r][c],a[r][c-1]);
                c--;
            }
            else if(a[r][c-1]=='#') continue;
            else{
                ll j=c-1;
                while(a[r][j]==']' || a[r][j]=='[') j--;
                if(a[r][j]=='#') continue;
                else{
                    for(int k=j;k<=c-1;k++) a[r][k]=a[r][k+1];
                    a[r][c]='.';
                    c--;
                }
            }
        }
        else if(moves[i]=='>'){
            if(a[r][c+1]=='.'){
                swap(a[r][c],a[r][c+1]);
                c++;
            }
            else if(a[r][c+1]=='#') continue;
            else{
                ll j=c+1;
                while(a[r][j]=='[' || a[r][j]==']') j++;
                if(a[r][j]=='#') continue;
                else{
                    for(int k=j;k>=c+1;k--) a[r][k]=a[r][k-1];
                    a[r][c]='.';
                    c++;
                }
            }
        }
        
        cout<<moves[i]<<endl<<i+1<<endl;
        for(int x=0;x<n;x++){
            for(int y=0;y<2*m;y++) cout<<a[x][y];
            cout<<endl;
        }
        cout<<endl;
        
    }
    // for(int x=0;x<n;x++){
    //         for(int y=0;y<2*m;y++) cout<<a[x][y];
    //         cout<<endl;
    //     }
    //     cout<<endl;

    ll ans=0;
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(a[i][j]=='[') ans+=(ll)100*i+j;
        }
    }
    cout<<ans;
    
}